protocol/bc/hash.go

```diff
@@ -1,103 +1,92 @@
 package bc
 
 import (
-	"bytes"
 	"database/sql/driver"
-	"encoding/hex"
-	"encoding/json"
-	"fmt"
+	"encoding/binary"
 	"io"
 
 	"golang.org/x/crypto/sha3"
 
 	"chain/crypto/sha3pool"
 	"chain/encoding/blockchain"
-	"chain/errors"
 )
 
-// Hash represents a 256-bit hash.  By convention, Hash objects are
-// typically passed as values, not as pointers.
-type Hash [32]byte
+// Hash represents a 256-bit hash.
 
-var EmptyStringHash = sha3.Sum256(nil)
+var EmptyStringHash = NewHash(sha3.Sum256(nil))
 
-// String returns the bytes of h encoded in hex.
-func (h Hash) String() string {
-	b, _ := h.MarshalText() // #nosec
-	return string(b)
+func NewHash(b [32]byte) (h Hash) {
+	h.FromByte32(b)
+	return h
+}
+
+func (h Hash) Byte32() (b32 [32]byte) {
+	binary.BigEndian.PutUint64(b32[0:8], h.V0)
+	binary.BigEndian.PutUint64(b32[8:16], h.V1)
+	binary.BigEndian.PutUint64(b32[16:24], h.V2)
+	binary.BigEndian.PutUint64(b32[24:32], h.V3)
+	return b32
+}
+
+func (h *Hash) FromByte32(b32 [32]byte) {
```

Comment:
Actually, I think probably we should just delete this method. Any call site that looks like

```go
h.FromByte32(b)
```

can be rewritten as one of

```go
h = NewHash(b) // or
*h = NewHash(b)
```